src/main/java/com/iota/iri/controllers/TransactionViewModel.java

```diff
@@ -976,66 +975,45 @@ public String toString() {
         Cache<Indexable, TransactionViewModel> cache = tangle.getCache(TransactionViewModel.class);
         if (cache.getSize() >= cache.getConfiguration().getMaxSize()) {
-            cacheRelease(tangle);
+            cacheRelease(tangle, cache);
         }
         cache.put(hash, transactionViewModel);
     }
 
     /**
      * Release all transactions from cache.
-     * 
+     *
      * @param tangle Tangle
      * @throws Exception Exception
      */
     public static void cacheReleaseAll(Tangle tangle) throws Exception {
         Cache<Indexable, TransactionViewModel> cache = tangle.getCache(TransactionViewModel.class);
-        Queue<Indexable> releaseQueueCopy = cache.getReleaseQueueCopy();
-        List<Indexable> hashesToRelease = new ArrayList<>();
-        List<Pair<Indexable, Persistable>> batch = new ArrayList<>();
-
-        while (!releaseQueueCopy.isEmpty()) {
-            Indexable hash = releaseQueueCopy.poll();
-            if (hash != null) {
-                TransactionViewModel tvm = cache.get(hash);
-                hashesToRelease.add(hash);
-                if (tvm != null && tvm.shouldPersist()) {
-                    tvm.setShouldPersist(false);
-                    cache.put(hash, tvm);
-                    batch.addAll(tvm.getSaveBatch());
-                }
-            }
+        Indexable hash = cache.getNextReleaseKey();
+        while(hash != null) {
+            releaseNext(tangle, cache, hash);
+            hash = cache.getNextReleaseKey();
         }
-
-        tangle.saveBatch(batch);
-        cache.release(hashesToRelease);
     }
 
     /**
-     * Release {@link CacheConfiguration#getReleaseCount()} items from the cache
-     * 
+     * Release the next item from the cache
+     *
      * @param tangle Tangle
      * @throws Exception Exception
      */
-    private static void cacheRelease(Tangle tangle) throws Exception {
-        Cache<Indexable, TransactionViewModel> cache = tangle.getCache(TransactionViewModel.class);
-        List<Pair<Indexable, Persistable>> batch = new ArrayList<>();
-        Queue<Indexable> releaseQueueCopy = cache.getReleaseQueueCopy();
-        List<Indexable> hashesToRelease = new ArrayList<>();
-
-        for (int i = 0; i < cache.getConfiguration().getReleaseCount(); i++) {
-            Indexable hash = releaseQueueCopy.poll();
-            if (hash != null) {
-                TransactionViewModel tvm = cache.get(hash);
-                hashesToRelease.add(hash);
-                if (tvm != null && tvm.shouldPersist()) {
-                    tvm.setShouldPersist(false);
-                    cache.put(hash, tvm);
-                    batch.addAll(tvm.getSaveBatch());
-                }
-            }
+    private static void cacheRelease(Tangle tangle, Cache<Indexable, TransactionViewModel> cache) throws Exception {
+        Indexable hash = cache.getNextReleaseKey();
```

Comment:
I would put this inside the implementation of the method below

Reply:
I agree, it doesn't look  like `cacheRelease()` is a necessary method, nor does it look like it's necessary to get the next release key, as the `cache.releaseNext()` call poll's the next releasable key anyways. So we're just fetching the key an extra time each time for no reason. You could instead set `cache.releaseNext()` to return a bool and that would let you check if the next hash is null by returning false from within the cache if it is. No more need for the `getNextReleaseKey()` that way.

Reply:
In order to persist and release, we need to get the next release key and check `shouldPersist` before.
I agree that we can use the `bool` return in the case of `cacheReleaseAll`